File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.py
from itertools import combinations

def solution(n, q, ans):
    list = [i for i in range(1, n + 1)]
    
    for i in range(len(ans)):
        if ans[i] == 0:
            for num in q[i]:
                if num in list:
                    list.remove(num)
    
    answer = 0
    
    for c in combinations(list, 5): 
        for i in range(len(q)):
            cnt = 0
            for j in q[i]:     
                if j in c:
                    cnt += 1
            if cnt != ans[i]:
                break  
        else:   
            answer += 1
    
    return answer

File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.java
import java.util.*;

class Solution {    
    List<Set<Integer>> combinations = new ArrayList<>();
    List<Integer> numList;
    int N;
    
    public int solution(int n, int[][] q, int[] ans) {
        N = ans.length;
        Set<Integer> possibleNum = new HashSet<>();
        
        for (int i = 1; i <= n; i++) {
            possibleNum.add(i);
        }
        
        for (int i = 0; i < N; i++) {
            if (ans[i] == 0) {
                for (int j = 0; j < 5; j++) {
                    possibleNum.remove(q[i][j]);
                }
            }
        }
        
        numList = new ArrayList<>(possibleNum);
        
        findCombinations(0, new HashSet<>());
        
        // System.out.println(combinations);
        
        int result = 0;
        
        for (Set<Integer> comb : combinations) {
            boolean possible = true;
            for (int i = 0; i < N; i++) {
                int count = 0;
                for (int j = 0; j < 5; j++) {
                    if (comb.contains(q[i][j]))
                        count++;
                }
                if (count != ans[i]) {
                    possible = false;
                    break;
                }
            }
            if (!possible)
                continue;
            else {
                result++;
            }
        }
        
        return result;   
    }
    
    public void findCombinations(int idx, Set<Integer> set) {
        if (set.size() == 5) {
            combinations.add(set);
            return;
        }
        
        for (int i = idx; i < numList.size(); i++) {
            Set<Integer> newSet = new HashSet<>(set);
            newSet.add(numList.get(i));
            findCombinations(i + 1, newSet);
        }
    }
}